js-packages/@prestojs/rest/src/InferredPaginator.ts

```diff
@@ -0,0 +1,230 @@
+import CursorPaginator, { CursorPaginationState } from './CursorPaginator';
+import { EndpointExecuteOptions } from './Endpoint';
+import LimitOffsetPaginator, { LimitOffsetPaginationState } from './LimitOffsetPaginator';
+import PageNumberPaginator, { PageNumberPaginationState } from './PageNumberPaginator';
+import { PaginationStateChangeListener, PaginatorInterface } from './Paginator';
+
+type PaginationState =
+    | CursorPaginationState
+    | PageNumberPaginationState
+    | LimitOffsetPaginationState;
+
+export default class InferredPaginator implements PaginatorInterface {
+    onChange?: PaginationStateChangeListener;
+    paginator: null | CursorPaginator | PageNumberPaginator | LimitOffsetPaginator = null;
+    initialState: Record<string, any>;
+
+    get currentState(): PaginationState | null | undefined {
+        return this.paginator?.currentState;
+    }
+
+    get total(): number | null | undefined {
+        if (this.paginator) {
+            if (this.paginator instanceof CursorPaginator) {
+                throw new Error('total not valid for cursor pagination');
+            }
```

Comment:
I was thinking that in most cases you know what pagination you are dealing with and so you'd write your UI for one type of pagination so I was erring on the side of letting user know about misuse rather than silently ignoring it.

For framework/generic code I would have thought you'd write the pagination UI as a component and switch it based on the inferred type, not based on the presence of each variable (eg. you'd have next/previous for cursor based but for a lot of page number based interfaces if they have next/previous at all they are to either side of pages)

eg.

```jsx
let paginationEl = null;

// If we know pagination type...
if (paginator.paginator) {
   if (paginator.paginator instanceof LimitOffsetPaginator) {
      paginationEl = <LimitOffsetPaginatorButtons paginator={paginator} />;
  }
   if (paginator.paginator instanceof PageNumberPaginator) {
      paginationEl = <PageNumberPaginationButtons paginator={paginator} />;
  }
   if (paginator.paginator instanceof CursorPaginator) {
      paginationEl = <CursorPaginatorButtons paginator={paginator} />;
  }
}
```

So I am tentatively on the side of current approach but can be convinced otherwise if you feel it's too strict